### ISSUE TITLE
Sync Folder?

### ISSUE THREAD

**Issue by Tiemichael** [haiwen/seadroid#502]
Is it possibe to sync specific folders?
It's working on the PC, but for android I am not able to find it.
Would be a great feature to keep e.g. important documents synchronized

**Comment by shoeper** [comment]
It is currently not possible in the android app.
A possible workaround is to use FolderSync with Seafiles WebDAV (Seafdav).

**Comment by Tiemichael** [comment]
Thanks!
Will try ....

**Comment by jult** [comment]
Or just use syncthing. At least over there somebody actually knows how to code.

**Comment by shoeper** [comment]
@jult if you don't like Seafile just keep away from it and follow syncthing instead.

**Comment by jult** [comment]
It's really strange, nobody on either side ever checked the other for code improvements, I've not seen any overlap whatsoever. Here's what I just posted at syncthing: https://github.com/syncthing/syncthing/issues/2970 and it seems to be contagious. Both packages are severely lacking each other's featureset. Merge the two and you'll have a killer app, one that kills all others, it could easily beat btsync, dropbox, tresorit, megasync, all of them.

**Comment by shoeper** [comment]
Do you know how Seafile works? I really doubt that syncthings code can efficiently be used to sync Seafile libraries.

**Comment by jult** [comment]
You do realize github is created by Linus Torvalds? He's being quite rude quite often, for a reason: It ends up giving results. https://www.youtube.com/watch?v=IVpOyKCNZYw or the entire interview: https://www.youtube.com/watch?v=MShbP3OpASA

**Comment by jult** [comment]
I'm using seafile for a while now, yes. It's basically the other end of the line; No config, no control. Using python.
Syncthing is all the rest, has what's missing in seafile. Should be what the "Advanced" button would show in seafile.

**Comment by shoeper** [comment]
You are not Linus. Linus did do (and still does) very very much for the OpenSource world.

And of course it would be great to be able to sync libraries but I'm pretty sure there is very much work involved. It should not just be able to sync but to do it as reliable as the desktop client does.

**Comment by jult** [comment]
Frankly, the whole "library" concept is not needed. Who really uses that name for anything? I only started seeing it appear in Windows 7 and the Windows Live world, and it annoyed the hell out of me. Virtual folders than just contain files/folders that suddenly need to be called "Libraries"? How to needlessly confuse the world.. Everybody I know that uses seafile or anything like seafile uses it to auto-sync folders, not "libraries". I do too. I pick drive-letters or folders that I would like to share with other devices, and auto-backup using seafile or syncthing. "My library" is just a top level folder. What it means to say is "Groups of folders and/or files", which is just simply a file-system, another folder that has subfolders. Or directories, which is perfectly fine to describe what seafile wants to do.

**Comment by shoeper** [comment]
> Frankly, the whole "library" concept is not needed. Who really uses that name for anything? I only started seeing it appear in Windows 7 and the Windows Live world, and it annoyed the hell out of me. Virtual folders that just contain files/folders that suddenly need to be called "Libraries"? How to needlessly confuse the world.. Everybody I know that uses seafile or anything like seafile uses it to auto-sync folders, not "libraries". I do too. I pick drive-letters or folders that I would like to share with other devices,

Technically this is not optimal (use of Seafile). Often data can be easily split into libraries.
And Seafile cannot remove the library concept. I admit that it was a bit strange on the first sight, but having a closer look at it libraries have many advantages. E.g. working with projects: One keeps the current project in sync and removes it locally afterwards while it can still be easily found on the server. Storage usage: One can adjust settings on per library basis, allowing to have unlimited history for documents but limited history for stuff being less important and changing regulary (=save storage). Furthermore the library concept allows Seafile to use git like sync which is the reason for the quite stable synchronisation including conflict handling.

**Comment by carlos22** [comment]
Folder Sync would be very useful for non-cloud apps like notes-taking, or todo.txt clients. Is there a reason why it is not implemented? Or just the usual, somebody needs to do it?

**Comment by rsramek** [comment]
I would also like to understand if I can expect this to come in future. I tried using seafile in 2014 for this very thing: syncing a directory of notes. I now came back to it but it turns out it's still not implemented :(

**Comment by freeplant** [comment]
I'm sorry we don't have a plan to implement folder sync on Android platform.

**Comment by carlos22** [comment]
A workaround is using the WebDAV (SeafDAV) interface to seafile and run any of the many (non libre) WebDAV Sync Tools. For Example: https://play.google.com/store/apps/details?id=dk.tacit.android.foldersync.lite&hl=de

Most of these apps can not access the SD Card on Android >= 6 - this also applies to Syncthing (https://syncthing.net/) which is another non-seafile (but free/libre) solution to sync two folders (but you could use it on a server in combination with seaf-fuse).

**Comment by rsramek** [comment]
Thanks for the direct response!

I think I will give syncthing a try once I go through it's documentation to understand what it does.

**Comment by jult** [comment]
Stopped using seafile altogether for a while now. syncthing does a fine job, has loads of improved features, and is able to sync exactly and only what I want, with the devices I want, and I run it on my own debian server for a per user instance using OpenMediaVault (see https://bintray.com/openmediavault-plugin-developers/erasmus-sync/openmediavault-syncthing ).

**Comment by laur89** [comment]
I'm another fool that jumped on the seafile bandwagon assuming the android app does what you'd expect from a syncing client.

@jult still recommending syncthing?

**Comment by jult** [comment]
@laur89 Actually, yes, I do still swear by syncthing. Very much so. As soon as you have it configged the way you want it to behave, it's running swiftly and just does what I want it to.
I have it on my android devices, it auto-syncs whenever I have it on charger and wifi (two great requisites by the way, brilliant thinking by the syncthing devs!), just feels super secure. And I use great ignore lists now, which help a lot; https://gist.github.com/jult/e2eaedad6b9e29d95977fea0ddffae7d
Plus, on my desktops I have it set to run with low CPU priority, or when cpu/storage is idle. It's just a treat how good it syncs with my NAS and partly with a remote server now. It's really been improved lots! The start is always a little hard, but I really like the fact that it can be configured so well and as much as you'd want..

**Comment by laur89** [comment]
Do you figure adding syncthing into the mix for syncing some seafile library subfolder would work?

**Comment by jult** [comment]
Of course, but I would not do that. Then you're using a lot of resources for nothing. I'd kill seafile, take some time to install and set up syncthing and you'll be a happy camper.

**Comment by nymilator** [comment]
@jult 
Can i use client side encryption with syncthing?
I search someting to backup some data from my android phone, but the backup should be encrypted

**Comment by canton7** [comment]
Syncthing doesn't have encrypted devices yet (data is encrypted in transit, but decrypted when it reaches a device). 

Tools like Syncthing are bad as backup tools though: if you accidentally delete a file, it automatically gets deleted from your backup as well.

**Comment by jult** [comment]
@canton7 Uh? You realize that is one of the most prominent settings to click on or off, to set a source or master, to allow it to delete the existing synced copies or not? Syncthing does versioning by default, so it keeps the files you've accidentally deleted. And by the way, there's not 1 backup-tool out there that does not do this by default. It's called 'mirroring', the standard de-facto way to backup using rsync --delete.

@nymilator As for encrypting stuff locally: You should consider *why* it's not doing that;
- If you're using any big OS of the day, the same open access syncthing has, is the one you have to the system when you're logged in, so either you encrypt your entire drive so it's invisible when you're offline, or you encrypt it on disk and can't access it while you have access to its passwords/keys anyway. If you can't access it, why do you have the copy?
You can and should sync to and from a TrueCrypt/[Cyphershed](https://en.wikipedia.org/wiki/CipherShed) volume. This is your safest bet right now.
That way, the synced copy is secure as soon as you leave or shutdown your system with the syncs on it. If your copy is on a NAS, same thing, when it's stolen or shutdown, your syncs are invisible, depending on how you open your TrueCrypt volume.
If your syncthing storage is on a remote untrusted device, you should ask yourself why you use that as a backup for anything: The untrusted device has access to the keys in traffic or on the device too, so why bother encrypting it on disk? Either way, seafile's trick for this is security by obscurity, so don't fall for that.

**Comment by canton7** [comment]
> Uh? You realize that is one of the most prominent settings to click on or off, to set a source or master, to allow it to delete the existing synced copies or not?

I think you're confusing several things:

 - There's the 'master' setting, which controls whether the "backup" device is allowed to change files, which are then propagated to the "master" device. This does not affect whether corruption/deletions are propagated to the backup straight away.
 - There's an advanced "ignoredeletes" setting, which prevents the "backup" device from ever deleting a file. This gets around the "deletion is propagated straight away", at the cost of accumulating deleted files forever.
 - There's versioning, which is a workaround, but still inferior to a proper backup tool.

**Comment by shoeper** [comment]
> Encrypting and Decrypting on the fly, especially if you want it to be really secure, is very heavy on CPU and RAM-usage.

This is not true anymore. Almost any modern CPU supports AES nowadays. E.g. my Laptop is LUKS encrypted, writing with 500 MiB/s has almost no CPU usage. AES performance is usually way better than disk performance even with multiple disks or SSDs.

**Comment by erenoglu** [comment]
So year 2018 and Seafile on Android still does not have proper sync. I'm also looking for an alternative. However, my seafile server is hosted on an untrusted  VPS and all files are locally encrypted than uploaded. So data on server is encrypted at rest. Can syncthing do this?  It's a required feature for my case.

**Comment by jult** [comment]
Just for the record; I've been forced to switch to nextcloud. It does miss some features of syncthing, but I can mitigate those because it can handle Web Distributed Authoring and Versioning (WebDAV) just fine, and there are perfect android apps for that. (The nextcloud app for android sucks for syncing.) Nextcloud also provides a lot of config and options, among them is to have disk-encrypted storage be the default. Still, I think in most cases this is a security by obscurity solution, since most people have non-stop access to the decrypted state of their files, which makes the thing somewhat moot, since whoever gains physical (thus usually also root) access to the system with encrypted files will also be able to see all traffic to and from the system, and the keys used for that traffic is easily captured. On the fly disk-encryption only makes sense if:
- You have way too much CPU and RAM doing nothing, and your storage I/O is really fast.
- You expect burglars to physically steal your storage media and don't want them to see what's on it, or some other "disconnect from network/stream" state would be expected or feared.

**Comment by r4dh4l** [comment]
> We don't have a plan to implement folder sync on Android platform yet.

Would it be possible to enhance the current one-directory syncing for pictures/videos to allow pushing other files to a library, for example podcasts downloaded on the smartphone?

**Comment by freeplant** [comment]
@r4dh4l This is possible. We will consider add such a feature if more users request.

**Comment by Gustl22** [comment]
@freeplant Are there any docs how the folder sync process is implemented in desktop client? So it would be easier to port it for volunteers.

**Comment by freeplant** [comment]
The sync process is very complicated in desktop client. It is not suitable for android.

**Comment by volkerkarle** [comment]
I'm using seafood to sync my zotero database. However, to use the zotero App, I need to find a way to sync specific folders on Android (which works fine on Linux). Any ideas?

**Comment by shoeper** [comment]
FolderSync (Pro)

**Comment by volkerkarle** [comment]
> FolderSync (Pro)

Can FolderSync also cope with Single Sign-On?

**Comment by shoeper** [comment]
You have to use WebDAV with FolderSync. I think there is a setting to set a webdav password when using sso.

**Comment by Gustl22** [comment]
I recommend CX file explorer as I got sync problems with FolderSync. Doesn't actually allow to save folders, but definitely is better at synchronising / latency / saving edited files directly.

**Comment by DenisMir** [comment]
Please implement this feature. It is 2026 and there is still no real two way sync on Android. So apps like Obsidian can't be used with Seafile. No the app framework is no option. Most apps don't support it.
